### PR TITLE
feat(optimizer): support column pruning for some operators

### PR DESCRIPTION
**Pull request by lokax**
Signed-off-by: lokax <m632656684@gmail.com>
#654

### CODE REVIEW

**Comment by skyzh** [comment]
That's so cool! I'll take a look later. cc @st1page may also help 🥰

**Review by xxchan — COMMENTED** [review]
I think it generally LGTM from the high level. But there are too many details in such a big PR which is hard to review... 🥹

File: src/optimizer/plan_nodes/logical_delete.rs [hunk 1]
```diff
@@ -46,6 +46,16 @@ impl PlanNode for LogicalDelete {
             false,
         )]
     }
+
+    fn prune_col(&self, _required_cols: BitSet) -> PlanRef {
+        let child_schema_len = self.child.out_types().len();
+        let mut input_cols = BitSet::with_capacity(child_schema_len);
+        (0..child_schema_len).into_iter().for_each(|col_idx| {
+            input_cols.insert(col_idx);
+        });
```
[the remark this answers is not in the document]
Comment (lokax) [reply src/optimizer/plan_nodes/logical_delete.rs:55]: I will modify it

File: src/optimizer/plan_nodes/logical_projection.rs [hunk 2]
```diff
@@ -106,6 +107,32 @@ impl PlanNode for LogicalProjection {
...
+
+        let mut visitor = CollectRequiredCols(BitSet::new());
+        new_projection_expressions
+            .iter()
+            .for_each(|expr| visitor.visit_expr(expr));
+
+        let input_cols = visitor.0;
+
+        if new_projection_expressions.is_empty() {
+            new_projection_expressions.push(BoundExpr::Constant(DataValue::Int32(81)));
```
[the remark this answers is not in the document]
Comment (lokax) [reply src/optimizer/plan_nodes/logical_projection.rs:125]: 81 is a magic number. 
If we have a existential subquery `SELECT v1 FROM v_table WHERE EXISTS (SELECT * FROM w_table)`, we don't need any column of w_table in the projection(expr: v1). After column pruning, the subquery projection expression len will become 0. But maybe we should a constant expression for upper operator in some subquery implementation. But for now , the 81 is not important.

File: src/optimizer/plan_nodes/dummy.rs [hunk 3]
```diff
@@ -8,10 +8,23 @@ use super::*;
 
 /// A dummy plan.
 #[derive(Debug, Clone, Serialize)]
-pub struct Dummy {}
+pub struct Dummy {
+    schema: Vec<ColumnDesc>,
```
[the remark this answers is not in the document]
Comment (lokax) [reply src/optimizer/plan_nodes/dummy.rs:12]: Maybe it was just a bug and I fixed it by the way
In SQL: `SELECT x FROM test WHERE 1 > 3`.
the projection operator `schema()` requires the schema of the following operator, `schema()` will panic if dummy schema hasn't schema.

File: src/optimizer/plan_nodes/logical_projection.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (xxchan) [reply src/optimizer/plan_nodes/logical_projection.rs:125]: If there's no output column, maybe we'd better refactor executor/chunk to allow such situation. Adding it here isn't correct in semantics: the `required_cols` isn't satisfied.

**Review by xxchan — APPROVED** [review]
Others generally LGTM. Let's merge it!

File: src/optimizer/plan_nodes/logical_projection.rs [hunk 4]
```diff
@@ -106,6 +107,32 @@ impl PlanNode for LogicalProjection {
...
+        let mut visitor = CollectRequiredCols(BitSet::with_capacity(required_cols.len()));
+        new_projection_expressions
+            .iter()
+            .for_each(|expr| visitor.visit_expr(expr));
+
+        let input_cols = visitor.0;
+
+        if new_projection_expressions.is_empty() {
+            new_projection_expressions.push(BoundExpr::Constant(DataValue::Int32(81)));
+        }
```
Comment (xxchan) [inline src/optimizer/plan_nodes/logical_projection.rs:126]:
```suggestion
```

**Pull request merged**